dvc/remote/ssh/__init__.py

```diff
@@ -278,3 +293,75 @@ def exists_with_progress(chunks):
                 ret = list(itertools.compress(checksums, in_remote))
                 pbar.update_desc("", 0)  # clear path name description
                 return ret
+
+    def link(self, from_info, to_info):
+        with self.ssh(from_info) as ssh:
+            self._link(from_info, to_info, self.cache_types, ssh)
+
+    def _link(self, from_info, to_info, link_types, ssh):
+        from_path = from_info.path
+        to_path = to_info.path
+
+        assert ssh.isfile(from_path)
+
+        dname = posixpath.dirname(to_path)
+
+        # XXX: We can remove the `exists` call, since `makedirs`
+        # will ignore it
+        if not ssh.exists(dname):
+            ssh.makedirs(dname)
+
+        # NOTE: just create an empty file for an empty cache
+        if ssh.getsize(from_path) == 0:
+            ssh.open(to_path, "w+").close()
+
+            msg = "Created empty file: {} -> {}".format(from_path, to_path)
+            logger.debug(msg)
+            return
+
+        self._try_links(from_info, to_info, link_types, ssh)
+
+    @classmethod
+    def _get_link_method(cls, link_type, ssh):
+        CACHE_TYPE_MAP = {"copy": ssh.cp, "symlink": ssh.symlink}
+
+        try:
+            return CACHE_TYPE_MAP[link_type]
+        except KeyError:
+            raise DvcException(
+                "Cache type: '{}' not supported!".format(link_type)
+            )
+
+    def _do_link(self, from_info, to_info, link_method, ssh):
+        # XXX: We are testing if file exists rather than if file is a link
+        if ssh.exists(to_info.path):
+            raise DvcException("Link '{}' already exists!".format(to_info))
+        else:
+            link_method(from_info.path, to_info.path)
+
+        if self.protected:
+            self.protect(to_info)
+
+        msg = "Created {}'{}': {} -> {}".format(
+            "protected " if self.protected else "",
+            self.cache_types[0],
+            from_info,
+            to_info,
+        )
+        logger.debug(msg)
+
+    def _try_links(self, from_info, to_info, link_types, ssh):
+        i = len(link_types)
+        while i > 0:
+            link_method = self._get_link_method(link_types[0], ssh)
+            try:
+                self._do_link(from_info, to_info, link_method, ssh)
+                return
+
+            except DvcException as exc:
+                msg = "Cache type '{}' is not supported: {}"
+                logger.debug(msg.format(link_types[0], str(exc)))
+                del link_types[0]
+                i -= 1
+
+        raise DvcException("no possible cache types left to try out.")
```

Comment:
The other issue is we pass `ssh` here everywhere. This could be solved by switching to transparent client/pool, e.g. an object, which will summon a client from a pool for each method call.

Reply:
@Suor , any example on how to use the _client/pool_ ? I prefer to use this instead of passing the connection everywhere

Reply:
@mroutis if you'll implement all those methods you won't need that I think. Just don't pass `ssh`, summon a connection every time you need.